spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/pubsub/it/PubSubChannelAdaptersIntegrationTests.java

```diff
@@ -201,6 +205,57 @@ public void sendAndReceiveMessageManualAck() {
 		});
 	}
 
+	// If this test flakes, delete it.
+	// It verifies that in AUTO_ACK mode, the message is neither acked nor nacked, and that
+	// redelivery happens after subscription's ackDeadline passes.
+	// There is also a client library bug (https://github.com/googleapis/java-pubsub/issues/141) that
+	// results in ackDeadline being extended by 60 seconds even when maxAckExtensionPeriod is zero,
+	// making minimum redelivery time is ackDeadline + 60.
+	@Test
+	public void sendAndReceiveMessageAutoAckWithFailure() {
+		ApplicationContextRunner customContextRunner = new ApplicationContextRunner()
+				.withConfiguration(AutoConfigurations.of(
+						GcpContextAutoConfiguration.class,
+						GcpPubSubAutoConfiguration.class))
+				.withUserConfiguration(SubscribableConfiguration.class, CommonConfiguration.class);
+
+		customContextRunner.run((context) -> {
+			try {
+				context.getBean(PubSubInboundChannelAdapter.class).setAckMode(AckMode.AUTO_ACK);
+				context.getBean("inputChannel", MessageChannel.class).send(
+						MessageBuilder.withPayload("This message is in trouble.".getBytes()).build());
+
+				SubscribableChannel channel = context.getBean("outputChannel", SubscribableChannel.class);
+
+				AtomicInteger numReceivedMessages = new AtomicInteger(0);
+				channel.subscribe(msg -> {
+					if (numReceivedMessages.incrementAndGet() == 1) {
+						throw new RuntimeException("BOOM!");
+					}
+				});
+
+				// wait for initial delivery
+				Awaitility.await().atMost(10, TimeUnit.SECONDS)
+						.until(() -> numReceivedMessages.get() > 0);
+				assertThat(numReceivedMessages.get()).isEqualTo(1);
+
+				// Expect redelivery within 2 minutes:
+				// 30 seconds subscription ackDeadline
+				// + 60 seconds https://github.com/googleapis/java-pubsub/issues/141
+				// + 30 seconds anti-flake buffer
+				Awaitility.await().atMost(120, TimeUnit.SECONDS)
+						.until(() -> numReceivedMessages.get() > 1);
+				assertThat(numReceivedMessages.get()).isEqualTo(2);
+			}
+			finally {
+				context.getBean(PubSubInboundChannelAdapter.class).stop();
+				PubSubAdmin pubSubAdmin = context.getBean(PubSubAdmin.class);
+				pubSubAdmin.deleteSubscription((String) context.getBean("subscriptionName"));
+				pubSubAdmin.deleteTopic((String) context.getBean("topicName"));
```

Comment:
Can we do this in `@After` this kind of code seems to be repeated across tests.

Reply:
Not the way it is right now, since it requires spring context to find the name of topic/subscription.
I can refactor the tests in a different PR, so that before/after are both done outside of the context.

Reply:
This would allow not relying on custom integration test configuration, I think. We could just rely on autoconfig.

Reply:
#2321 to track